index.js

```diff
@@ -1,24 +1,70 @@
 const Router = require('express').Router;
 
-module.exports = (options) => {
+/**
+ * A helper for `createNamedAsyncMiddleware`
+ * If the function already has a name, we'll just use that and tack
+ * `Async` on the end. If the function doesn't have a name we'll
+ * use the path for the target handler plus the index of the current
+ * middleware.
+ */
+function getFunctionName(fun, handlerPath, middlewareIndex) {
+  if (fun.name && fun.name !== 'anonymous') {
+    return fun.name + 'Async';
+  }
+
+  // Path can be a string, regex, or array of either
+  const strPath = handlerPath.toString();
+
+  return strPath + middlewareIndex;
+}
+
+/**
+ * Create a middleware which calls the original function but catches
+ * unhandled promise rejections (either from async functions or returned promise)
+ * and passes the error to `next` which will call the next error handler middleware.
+ *
+ * For better stack traces we want to make sure that the functions are
+ * named function for each middleware the simplest to have a dynamic
+ * function.name is to initialize it in an object with the key being
+ * the dynamic function name
+ */
+function createNamedAsyncMiddleware(fun, handlerPath, middlewareIndex) {
+  const functionName = getFunctionName(fun, handlerPath, middlewareIndex);
+
+  return {
+    [functionName]: (req, res, next) => {
+      // Execute the function
+      const p = fun(req, res, next);
+
+      // If the result isn't a promise
+      if (!p || typeof p !== 'object' || !p.then || !p.catch) {
+        console.error(
+          'Expected then middleware number %d for route %s to return a promise',
+          i,
+          req.url,
+        );
+
+        return;
+      }
+
+      // catch the error and pass it to the next error handling middleware
+      const pp = p.catch(next);
+      if (pp.done) {
+        // Terminate promise for bluebird and Q
+        pp.done();
+      }
+    },
+  }[functionName];
```

Comment:
why put this in an object and then get the key out by that object? Is it to shim in a function name?

Reply:
Yeah, I'm unaware of any other way to set a function name dynamically.

Reply:
The comment explains